File: fix_encoding.js
// Fix triple/double encoded UTF-8 Arabic text in HTML files
const fs = require('fs');

function fixMojibake(text) {
    // Pattern: sequences of chars in range \x80-\xFF that form valid UTF-8 when converted back to bytes
    let result = '';
    let i = 0;
    
    while (i < text.length) {
        const code = text.charCodeAt(i);
        
        if (code < 0x80) {
            // Plain ASCII - keep as-is
            result += text[i];
            i++;
        } else if (code >= 0x80 && code <= 0xFF) {
            // Potential mojibake sequence - collect consecutive chars in 0x80-0xFF range
            let byteArr = [];
            let j = i;
            while (j < text.length && text.charCodeAt(j) >= 0x80 && text.charCodeAt(j) <= 0xFF) {
                byteArr.push(text.charCodeAt(j));
                j++;
            }
            
            // Try to decode the byte sequence as UTF-8
            const buf = Buffer.from(byteArr);
            try {
                const decoded = buf.toString('utf8');
                // Check if decoded is valid (no replacement chars)
                if (!decoded.includes('\uFFFD')) {
                    result += decoded;
                } else {
                    // Can't fix, keep original
                    result += text.substring(i, j);
                }
            } catch(e) {
                result += text.substring(i, j);
            }
            i = j;
        } else {
            // Higher unicode - already proper Unicode char
            result += text[i];
            i++;
        }
    }
    return result;
}

function processFile(filepath) {
    console.log('Processing:', filepath);
    
    // Read file as binary buffer, then decode as latin-1 (1-to-1 mapping)
    const raw = fs.readFileSync(filepath);
    // Convert bytes to string using latin-1 (each byte maps to same code point)
    let text = '';
    for (let i = 0; i < raw.length; i++) {
        text += String.fromCharCode(raw[i]);
    }
    
    // Fix the mojibake
    const fixed = fixMojibake(text);
    
    // Write back as UTF-8
    fs.writeFileSync(filepath, fixed, 'utf8');
    console.log('  Saved!');
}

const files = [
    'C:\\My Web Sites\\test\\tallo-menu\\menu-ar.html',
    'C:\\My Web Sites\\test\\tallo-menu\\menu-en.html',
];

for (const f of files) {
    if (fs.existsSync(f)) {
        try {
            processFile(f);
        } catch(e) {
            console.error('  ERROR:', e.message);
        }
    } else {
        console.log('NOT FOUND:', f);
    }
}

console.log('All done!');

File: fix_encoding.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Fix double-encoded UTF-8 Arabic text in HTML files.
The original HTTrack-downloaded file had Arabic UTF-8 bytes
that were read as Latin-1 and then re-encoded as UTF-8 (mojibake).
"""
import os
import re

def fix_mojibake(text):
    """
    Convert mojibake text back to proper Arabic.
    Each mojibake char is <= 0xFF and originally a UTF-8 byte.
    We collect these chars and convert them back to proper UTF-8.
    """
    result = []
    i = 0
    chars = list(text)
    n = len(chars)
    
    while i < n:
        c = chars[i]
        code = ord(c)
        
        # ASCII range - keep as is
        if code < 0x80:
            result.append(c)
            i += 1
        # These are the double-encoded Arabic chars (all <= 0xFF)
        elif code <= 0xFF:
            # Collect consecutive non-ASCII chars that are <= 0xFF
            byte_seq = []
            while i < n and 0x80 <= ord(chars[i]) <= 0xFF:
                byte_seq.append(ord(chars[i]))
                i += 1
            # Try to decode as UTF-8
            try:
                arabic = bytes(byte_seq).decode('utf-8')
                result.append(arabic)
            except UnicodeDecodeError:
                # If can't decode, keep as is
                for b in byte_seq:
                    result.append(chr(b))
        else:
            # Already proper Unicode (>0xFF), keep as is
            result.append(c)
            i += 1
    
    return ''.join(result)

def process_file(filepath):
    print(f"Processing: {filepath}")
    
    with open(filepath, 'rb') as f:
        raw = f.read()
    
    # Decode as UTF-8
    try:
        text = raw.decode('utf-8')
    except UnicodeDecodeError:
        print(f"  ERROR: Cannot decode as UTF-8")
        return
    
    # Check if there's mojibake (chars in range 0x80-0xFF that form UTF-8 sequences)
    mojibake_pattern = re.compile(r'[\xc0-\xff][\x80-\xbf]+')
    if not mojibake_pattern.search(text):
        print(f"  No mojibake found, skipping.")
        return
    
    fixed = fix_mojibake(text)
    
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(fixed)
    
    print(f"  Fixed and saved!")

files = [
    r'C:\My Web Sites\test\tallo-menu\menu-ar.html',
    r'C:\My Web Sites\test\tallo-menu\menu-en.html',
    r'C:\My Web Sites\test\tallo-menu\index.html',
    r'C:\My Web Sites\test\tallo-menu\opinion.html',
    r'C:\My Web Sites\test\tallo-menu\opinion-en.html',
    r'C:\My Web Sites\test\tallo-menu\css\style-1.8.css',
]

for f in files:
    if os.path.exists(f):
        process_file(f)
    else:
        print(f"NOT FOUND: {f}")

print("Done!")
